src/sqlfluff/rules/L003.py

```diff
@@ -242,30 +338,66 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
                 "hanging_lines": [],
                 # comment_lines keeps track of lines which are all comment.
                 "comment_lines": [],
+                # segments we've seen the last child of
+                "finished": set(),
+                # First non-whitespace node on a line.
+                "trigger": None,
             }
 
         if segment.is_type("newline"):
             memory["in_indent"] = True
-            # We're not going to flag on empty lines so we can safely proceed
-            return LintResult(memory=memory)
         elif memory["in_indent"]:
             if segment.is_type("whitespace"):
                 # it's whitespace, carry on
-                return LintResult(memory=memory)
+                pass
             elif segment.segments or (segment.is_meta and segment.indent_val != 0):
                 # it's not a raw segment or placeholder. Carry on.
-                return LintResult(memory=memory)
+                pass
             else:
                 memory["in_indent"] = False
                 # we're found a non-whitespace element. This is our trigger,
                 # which we'll handle after this if-statement
+                memory["trigger"] = segment
         else:
             # Not in indent and not a newline, don't trigger here.
+            pass
+
+        # Is this the last segment? If so, need to "flush" any leftovers.
+        is_last = self._is_last_segment(segment, memory, parent_stack, siblings_post)
+
+        if not segment.is_type("newline") and not is_last:
+            # We only process complete lines or on the very last segment
+            # (since there may not be a newline on the very last line)..
             return LintResult(memory=memory)
 
+        if raw_stack and raw_stack[-1] is not segment:
+            raw_stack = raw_stack + (segment,)
         res = self._process_raw_stack(
-            raw_stack + (segment,), tab_space_size=self.tab_space_size
+            raw_stack,
+            memory,
+            tab_space_size=self.tab_space_size,
+            templated_file=templated_file,
         )
+
+        if res:
+            # Saw a newline or end of parse tree. Is the current line empty?
+            trigger_segment = memory["trigger"]
+            if trigger_segment:
+                # Not empty. Process it.
+                result = self._process_current_line(res, memory)
+                if segment.is_type("newline"):
+                    memory["trigger"] = None
+                return result
+        return LintResult(memory=memory)
+
+    def _process_current_line(self, res, memory):
```

Comment:
I split part of the `_eval()` function into this new function to make this rule easier to understand. Not technically necessary.